consensus/state.go

```diff
@@ -1079,19 +1079,37 @@ func (cs *State) enterNewRound(height int64, round int32) {
 	}
 }
 
-// needProofBlock returns true on the first height (so the genesis app hash is signed right away)
-// and where the last block (height-1) caused the app hash to change
+// needProofBlock returns true if additional proof block needs to be created.
+// It happens on the first height (so the genesis app hash is signed right away)
+// and where the last block (height-1) or the previous one (height - 2) caused
+// the app hash to change.
 func (cs *State) needProofBlock(height int64) bool {
 	if height == cs.state.InitialHeight {
 		return true
 	}
 
-	lastBlockMeta := cs.blockStore.LoadBlockMeta(height - 1)
-	if lastBlockMeta == nil {
+	h1BlockMeta := cs.blockStore.LoadBlockMeta(height - 1)
+	if h1BlockMeta == nil {
 		panic(fmt.Sprintf("needProofBlock: last block meta for height %d not found", height-1))
 	}
+	h1Equal := bytes.Equal(cs.state.AppHash, h1BlockMeta.Header.AppHash)
```

Comment:
I would return immediately if h1 is not equal as there's no reason to load block at h-2.